libkbfs/block_tree.go

```diff
@@ -291,159 +311,214 @@ func (bt *blockTree) getBlocksForOffsetRange(ctx context.Context,
 	prefixOk bool, getDirect bool) (pathsFromRoot [][]parentBlockAndChildIndex,
 	blocks map[BlockPointer]Block, nextBlockOffset Offset,
 	err error) {
-	nextBlockOffset = pblock.FirstOffset()
-	if !pblock.IsIndirect() {
-		// Return a single empty path, under the assumption that the
-		// caller already checked the range for this block.
-		if getDirect {
-			// Return a child map with only this block in it.
-			return [][]parentBlockAndChildIndex{nil},
-				map[BlockPointer]Block{ptr: pblock}, nil, nil
+
+	// Make a queue for tasks and one for results. Queue for results is big to prevent workers from blocking on
+	// sending back their results.
+	jobs := make(chan task, 2)
+	results := make(chan result, 400)
+
+	// make a WaitGroup to keep track of whether there's still work to be done.
+	// TODO: simplify this so we don't need both the errgroup and the WaitGroup
```

Comment:
Yeah I think we can simplify all this a bit, and just use a `WaitGroup`:

1. Add an `error` field to `result`.
2. Make a new context using `context.WithCancel()`, and call `defer cancel()`.
3. Use `go` to launch your 16 workers, passing each of them the new context.
4. Use `wg` to track the length of the jobs queue, as you already do below.
5. If any of the results have a non-nil error (subject to the `DeadlineExceeded` check), just return it immediately, which will cancel the context via the `defer` and cause each of the workers will exit.

It's kind of like implementing your own `errgroup`, except now each error is tied precisely to the job that triggered it.  It wouldn't be a huge improvement, just a little one.  I'm open to other suggestions, if you already had some other simplification in mind.